scripts/add-celltype-fields-scpca-meta.py

```diff
@@ -0,0 +1,131 @@
+project_df = pandas.read_csv(
+    args.project_celltype_file, sep="\t", keep_default_na=False
+)
+
+# remove any extra / at the end
+bucket = args.bucket.strip("/")
+checkpoints_prefix = args.checkpoints_prefix.strip("/")
+
+# join library and project cell type metadata
+all_metadata_df = library_df.join(
+    project_df.set_index("scpca_project_id"), on="scpca_project_id", how="left"
+)
+# need to filter to only projects that have at least one reference (either SingleR or CellAssign)
+all_metadata_df = all_metadata_df[
+    all_metadata_df[["singler_ref_file", "cellassign_ref_file"]].notnull().all(1)
+]
+
+# remove any extra / at the end
+bucket = args.bucket.strip("/")
+checkpoints_prefix = args.checkpoints_prefix.strip("/")
+
+# set up S3
+s3 = boto3.resource("s3")
+s3_bucket = s3.Bucket(bucket)
+
+# go through every run id and modify scpca-meta.json files if present
+for run in all_metadata_df.itertuples():
+    print(f"Processing {run.scpca_run_id}")
+
+    # build paths to singleR and CellAssign checkpoints
+    celltype_checkpoints = {
+        "singler_checkpoint": f"{checkpoints_prefix}/celltype/{run.scpca_library_id}/{run.scpca_library_id}_singler/scpca-meta.json",
+        "cellassign_checkpoint": f"{checkpoints_prefix}/celltype/{run.scpca_library_id}/{run.scpca_library_id}_cellassign/scpca-meta.json",
+    }
+
+    # cell type fields that need to be updated
+    new_fields = {
+        "singler_model_file": f"s3://scpca-references/celltype/singler_models/{run.singler_ref_file}",
+        "cellassign_reference_file": f"s3://scpca-references/celltype/cellassign_references/{run.cellassign_ref_file}",
+    }
+
+    # for each checkpoint file, add cellassign and singler file path
+    for checkpoint_file in celltype_checkpoints.values():
+        # make sure file exists, otherwise just skip
+        try:
+            checkpoint_obj = (
+                s3_bucket.Object(checkpoint_file).get()["Body"].read().decode("utf-8")
+            )
+            checkpoint_meta = json.loads(checkpoint_obj)
+        except s3.meta.client.exceptions.NoSuchKey:
+            print(f"No cell type scpca-meta.json file for {run.scpca_run_id}")
+            continue
+
+        # only modify if existing fields don't match what's in the project metadata
+        if (
+            checkpoint_meta["singler_model_file"] == new_fields["singler_model_file"]
+            and checkpoint_meta["cellassign_reference_file"]
+            == new_fields["cellassign_reference_file"]
+        ):
+            print(
+                f"All fields are present, no updates for {run.scpca_run_id} to {checkpoint_file}"
+            )
+        else:
+            # before modifying, make sure that the entries in the project metadata aren't NA
+            if run.singler_ref_file != "NA":
+                checkpoint_meta["singler_model_file"] = new_fields["singler_model_file"]
+            else:
+                checkpoint_meta["singler_model_file"] = "NA"
```

Comment:
Reversing the logic here makes it a bit more clear?
```suggestion
            if run.singler_ref_file == "NA":
                checkpoint_meta["singler_model_file"] = "NA"
            else:
                checkpoint_meta["singler_model_file"] = new_fields["singler_model_file"]
```

Reply:
Related to my first comment: I think we actually want to be more careful here, and only update if the old file had the old format. If the file name includes a version string, I think we want to _not_ update it, to save us from ourselves in the future. We should not need to run this in the future, but if we do, we don't want to accidentally change any version strings.